### DISCUSSION TITLE
Updating OneToMany relationship on item loaded without populate removed all other relationships

### DISCUSSION THREAD

**Discussion by WingmanImd** [mikro-orm/mikro-orm#5883] [General]
### Describe the bug

Updating an OneToMany relationship on a model where the relationship isn't populated deletes all other items relationships.

### Reproduction

```
@Entity({ tableName: 'users' })
export class User {
    @PrimaryKey()
    id!: number;
    
    @ManyToOne({ entity: () => Role, nullable: true })
    role!: Rel<Role>;
}

@Entity({ tableName: 'roles' })
export class Role {
    @PrimaryKey()
    id!: number;
    @OneToMany(() => User, user => user.role)
    users = new Collection<User>(this);
}
```

```
const role = await em.findOne(Role, roleId);
const users = await em.find(User, { id: { $in: [2]} });
role.users.set(users)
```

^ this executes the following deleting the role_id from all other users
```
[query] update `users` set `role_id` = 11, `updated_at` = '2024-07-31 11:53:30.138' where `id` = 2 [took 53 ms, 1 row affected]

// this shouldn't run
[query] update `users` set `role_id` = NULL where (`id`) not in ((2)) [took 52 ms]
```

Populating users fixes the issue
```
const role = await em.findOne(Role, roleId, { 
   populate: ['users']
});

```

### What driver are you using?

@mikro-orm/mariadb

### MikroORM version

6.3.1

### Node.js version

20.10.0

### Operating system

Windows 10

### Validations

- [X] Read the [Contributing Guidelines](https://github.com/mikro-orm/mikro-orm/blob/master/CONTRIBUTING.md).
- [X] Read the [docs](https://mikro-orm.io/docs).
- [X] Check that there isn't [already an issue](https://github.com/mikro-orm/mikro-orm/issues) that reports the same bug to avoid creating a duplicate.
- [X] Check that this is a concrete bug. For Q&A open a [GitHub Discussion](https://github.com/mikro-orm/mikro-orm/discussions) or join our [Discord](https://discord.gg/w8bjxFHS7X).
- [X] The provided reproduction is a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of the bug.

**Comment by B4nan** [comment]
I don't see anything wrong in there, you call `set` on a collection, which means replace all items with what you provide. If the collection is not populated, it will run the query you don't like to ensure there are no items connected to it, only those you are setting there, if you populate it first, the ORM will compute a diff and only fire queries that are required.

Calling `set()` is the same as if you call `removeAll()` followed by `add()`, you override the collection state completely.

**Reply by WingmanImd** [reply]
"Calling set() is the same as if you call removeAll() followed by add(), you override the collection state completely."
That would be the behaviour I would expect. But the problem here is that the role.users.set([x,y,z]) deletes(sets to NULL) the role from all the users in the database not only those that where linked to the role.

Or do I have a wrong understanding of what happens behind the hood and what I think it's an "array" of users linked to a specific "role" is just and empty collection of users, and when I call set it just wipes all roles from all users

**Reply by B4nan** [reply]
Oh now I see what you mean, the query should be limited to the old FK value, like this:

```sql
update `users` set `role_id` = NULL where (`id`) not in ((2)) and `role_id` = 11
```

Will try to look into this later today.

**Reply by WingmanImd** [reply]
Exactly. Seems like I didn't point that out well enough :) Thanks!

**Reply by B4nan** [reply]
This should help 517108bd5a2defd15a2b6d677429e0b4d06d5422
